new-packages/Core/src/Ranges/Progressions.candy

```diff
@@ -0,0 +1,114 @@
+use ...Bool
+use ...Collections
+use ...Int
+use ...Maybe
+use ...Operators
+
+public trait IntProgression: Equals {
+  fun startBound(): Maybe[Int]
+  fun endBound(): Maybe[Int]
+  fun stepSize(): Int
+
+  fun contains(item: Int): Bool {
+    let startMatches = this.startBound()
+      .map[Bool]({ it <= item & (item - it) % this.stepSize() == 0 })
+      .else({ true() })
+    let endMatches = this.endBound()
+      .map[Bool]({ item <= it & (it - item) % this.stepSize() == 0 })
+      .else({ true() })
+    startMatches & endMatches
+  }
+}
+
+
+public type BoundedIntProgression = (start: Int, end: Int, step: Int)
+# A `Progression` where `this.start` is inclusive and `this.end` is exclusive.
+#
+# Invariants:
+# * `this step != 0`
+# * `(this step > 0) then { this end > this start } else { this end < this start }`
+# * `(this end - 1 - this start) % this step == 0`
```

Comment:
Is this what we want? Should `BoundedIntProgression(step = 3, start = 0, end = 4)` be valid?

Reply:
Yes, that should be what we discussed before. Or do you think we should encode progressions differently?

Reply:
I somehow thought we had the end as an exclusive range, but it's `step` larger than the last yielded value. So, for the progression `0, 3, 6, 9` it would be `step = 3, start = 0, end = 12`. So, basically, `(end - start) % step == 0`, and `(end - start) / step` is the number of items in the progression.